--- a/modules/protocol/hybrid.cpp	2026-02-15 05:54:45.029498497 +0100
+++ b/modules/protocol/hybrid.cpp	2026-02-15 05:35:03.708091000 +0100
@@ -15,11 +15,13 @@
 
 #include "module.h"
 #include "modules/chanserv/mode.h"
+#include "modules/nickserv/sasl.h"
 
 static Anope::string UplinkSID;
 
 class HybridProto final
 	: public IRCDProto
+	, SASL::ProtocolInterface
 {
 	void SendSVSKill(const MessageSource &source, User *u, const Anope::string &buf) override
 	{
@@ -28,7 +30,7 @@
 	}
 
 public:
-	HybridProto(Module *creator) : IRCDProto(creator, "ircd-hybrid 8.2.34+")
+	HybridProto(Module *creator) : IRCDProto(creator, "ircd-hybrid 8.2.34+"), SASL::ProtocolInterface(creator)
 	{
 		DefaultPseudoclientModes = "+oi";
 		CanSVSNick = true;
@@ -270,6 +272,29 @@
 		Uplink::Send("SVSHOST", u->GetUID(), u->timestamp, u->host);
 	}
 
+	/* SASL protocol interface methods */
+	void SendSASLMessage(const SASL::Message &message) override
+	{
+		Server *s = Server::Find(message.target.substr(0, 3));
+		auto target = s ? s->GetName() : message.target.substr(0, 3);
+
+		auto newparams = message.data;
+		newparams.insert(newparams.begin(), { target, "SASL", message.source, message.target, message.type });
+		Uplink::SendInternal({}, Me, "ENCAP", newparams);
+	}
+
+	void SendSVSLogin(const Anope::string &uid, NickAlias *na) override
+	{
+		Server *s = Server::Find(uid.substr(0, 3));
+		Anope::string target = s ? s->GetName() : uid.substr(0, 3);
+
+		if (na)
+		{
+			Uplink::Send("ENCAP", target, "SVSLOGIN", uid, '*', '*',
+				na->GetVHostHost().empty() ? "*" : na->GetVHostHost(), na->nc->display);
+		}
+	}
+
 	bool IsExtbanValid(const Anope::string &mask) override
 	{
 		return mask.length() >= 4 && mask[0] == '$' && mask[2] == ':';
@@ -649,6 +674,28 @@
 	}
 };
 
+struct IRCDMessageEncap final
+	: IRCDMessage
+{
+	IRCDMessageEncap(Module *creator) : IRCDMessage(creator, "ENCAP", 4) { SetFlag(FLAG_REQUIRE_SERVER); SetFlag(FLAG_SOFT_LIMIT); }
+
+	/*                                0 1    2         3         4    5      */
+	/* :0MC ENCAP services.chatik.pl  * SASL 0MCAAAAAB *         S    PLAIN  */
+	/* :0MC ENCAP services.chatik.pl  * SASL 0MCAAAAAB 0MCAAAAAC C    base64 */
+	void Run(MessageSource &source, const std::vector<Anope::string> &params, const Anope::map<Anope::string> &tags) override
+	{
+		if (params[1] == "SASL" && SASL::service && params.size() >= 6)
+		{
+			SASL::Message m;
+			m.source = params[2];
+			m.target = params[3];
+			m.type = params[4];
+			m.data.assign(params.begin() + 5, params.end());
+			SASL::service->ProcessMessage(m);
+		}
+	}
+};
+
 class ProtoHybrid final
 	: public Module
 {
@@ -677,6 +724,7 @@
 	/* Our message handlers */
 	IRCDMessageBMask message_bmask;
 	IRCDMessageCapab message_capab;
+	IRCDMessageEncap message_encap;
 	IRCDMessageCertFP message_certfp;
 	IRCDMessageEOB message_eob;
 	IRCDMessageJoin message_join;
@@ -774,6 +822,7 @@
 		message_whois(this),
 		message_bmask(this),
 		message_capab(this),
+		message_encap(this),
 		message_certfp(this),
 		message_eob(this),
 		message_join(this),
